EU4toV2/Source/V2World/Diplomacy/Diplomacy.cpp

```diff
@@ -63,46 +68,106 @@ void V2::Diplomacy::convertDiplomacy(
-			agreement.getAgreementType() == "client_vassal" || 
-			agreement.getAgreementType() == "daimyo_vassal" || 
-			agreement.getAgreementType() == "protectorate" || 
-			agreement.getAgreementType() == "tributary_state" || 
-			agreement.getAgreementType() == "march" || 
-			agreement.getAgreementType() == "colonial" ||
-			agreement.getAgreementType() == "colony" ||
-			agreement.getAgreementType() == "union" ||
-			agreement.getAgreementType() == "personal_union")
+		if (agreementMapper.isAgreementInDoublesiders(agreement.getAgreementType()))
+		{
+			// doublesiders are bidirectional; influence level +1, but never exceed 4
+			// They don't set military access, as it's not implied (not even for alliances).
+			if (r1.getLevel() < 4) r1.setLevel(r1.getLevel() + 1);
+			r1.increaseRelations(50);
+			if (r2.getLevel() < 4) r2.setLevel(r2.getLevel() + 1);
+			r2.increaseRelations(50);
+		}
+
+		if (agreementMapper.isAgreementInTributaries(agreement.getAgreementType()))
+		{
+			// influence level 5 - sphere, but not vassal, and military access is implied.
+			r1.setLevel(5);
+			r1.increaseRelations(75);
+			r1.setInfluence(20);
+			r1.setAccess(true);
+			r2.setAccess(true);
+		}
+
+		if (agreementMapper.isAgreementInVassals(agreement.getAgreementType()))
 		{
 			// Yeah, we don't do marches, clients or all that. Or personal unions. PUs are a second relation
 			// beside existing vassal relation specifying when vassalage ends.
-			// We assume all rulers are CK2 immortals and vassalage does not end with a specific date.
+			// However, vanilla Vic2 has PU end dates based on historical events, and we don't simulate those (for now).
 			agreement.setAgreementType("vassal");
 			r1.setLevel(5);
-			country2->second->addPrestige(-country2->second->getPrestige());
+			r1.increaseRelations(75);
+			r1.setInfluence(50);
+			// In vic2 military access through vassals is not automatic but is implied.
+			r1.setAccess(true);
+			r2.setAccess(true);
+
+			// We need to calculate devs and conglomerate devs (vassals + overlord) so that we can alter starting prestige
+			// of individual vassals. We cannot do so yet as we don't know who's alive, dead or a vassal at all.
+			if (!vassalCache[V2Tag2]) vassalCache[V2Tag2] = country2->second->getSourceCountry()->getTotalDev();
+			if (!masterCache[V2Tag1]) masterCache[V2Tag1] = country1->second->getSourceCountry()->getTotalDev();
+			masterVassals[V2Tag1].insert(V2Tag2);
 		}
 
-		// In essence we only have 3 diplomacy categories and these are it.
+		// In essence we should only recognize 3 diplomacy categories and these are it.
 		if (agreement.getAgreementType() == "alliance" || agreement.getAgreementType() == "vassal" || agreement.getAgreementType() == "guarantee")
 		{
 			// copy agreement
 			Agreement v2agreement(V2Tag1, V2Tag2, agreement.getAgreementType(), agreement.getStartDate());
 			agreements.push_back(v2agreement);
 		}
 	}
+	// Now, alter prestige of vassals.
```

Comment:
It won't let me comment above, but convertDiplomacy() is really long and doing several discrete tasks at multiple levels of abstraction. Most or all of the work should be put into functions called by convertDiplomacy()

Reply:
done, we have 5-6 new functions now coming from in there dealing with discrete tasks.